lib/evmone/eof.cpp

```diff
@@ -34,51 +61,85 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
 
     auto state = State::section_id;
     uint8_t section_id = 0;
+    uint16_t section_num = 0;
     EOFSectionHeaders section_headers{};
     const auto container_end = container.end();
     auto it = container.begin() + std::size(MAGIC) + 1;  // MAGIC + VERSION
+    // TODO: Since all sections are mandatory and they have to be ordered (Types, Code+, Data)
+    // TODO: this fragment of code can be much simpler. Rewriting needed.
+    uint8_t expected_section_id = TYPE_SECTION;
     while (it != container_end && state != State::terminated)
     {
         switch (state)
         {
         case State::section_id:
         {
             section_id = *it++;
+
+            if (section_id != expected_section_id)
+                return get_section_missing_error(expected_section_id);
+
             switch (section_id)
             {
             case TERMINATOR:
-                if (section_headers[CODE_SECTION] == 0)
-                    return EOFValidationError::code_section_missing;
                 state = State::terminated;
                 break;
-            case DATA_SECTION:
-                if (section_headers[CODE_SECTION] == 0)
-                    return EOFValidationError::code_section_missing;
-                if (section_headers[DATA_SECTION] != 0)
-                    return EOFValidationError::multiple_data_sections;
+            case TYPE_SECTION:
+                expected_section_id = CODE_SECTION;
                 state = State::section_size;
                 break;
             case CODE_SECTION:
-                if (section_headers[CODE_SECTION] != 0)
-                    return EOFValidationError::multiple_code_sections;
+            {
+                if (it >= container_end - 2)
```

Comment:
`it` here points to the next byte after section_id, I think the right condition here is
```suggestion
                if (it + 1 >= container_end)
```

Reply:
Fixed.